### PR TITLE
REGRESSION: [language filter plugin] 301 redirect without caching in browser

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/11159 and https://github.com/joomla/joomla-cms/issues/11143.
#### Summary of Changes

Adds HTTP headers to not cache in browsers the language filter plugin 301 redirect.

The problem was created by https://github.com/joomla/joomla-cms/pull/8894.
#### Testing Instructions

See https://github.com/joomla/joomla-cms/issues/11159 and https://github.com/joomla/joomla-cms/issues/11143.

pinging @infograf768 and @ggppdk

### CODE REVIEW

File: plugins/system/languagefilter/languagefilter.php [hunk 1]
```diff
@@ -378,6 +378,10 @@ public function parseRule(&$router, &$uri)
 				}
 			}
 
+			$this->app->setHeader('Cache-Control', 'no-store, no-cache, must-revalidate, post-check=0, pre-check=0');
+			$this->app->setHeader('Expires', 'Mon, 26 Jul 1997 05:00:00 GMT');
```
Comment (brianteeman) [inline plugins/system/languagefilter/languagefilter.php]: If there is no specific reason for that exact date can we make it the same as https://github.com/joomla/joomla-cms/blob/4a218fe3056b61dcd571a6ac9419c42d7be98329/libraries/joomla/application/web.php#L440

Comment (andrepereiradasilva) [inline plugins/system/languagefilter/languagefilter.php]: done

**Comment by ggppdk** [comment]
Confirmed that in both cases the 301 redirect for home-page without language code
- is not cached for browsers: chrome, firefox, edge

**Issue** #11159  ("Remove language code" for default language is ON (lang-filter plugin)) 
- language switching for all languages while in home page is now possible
- type or click the home page without language code then he will get the default language without any redirect

This is expected and correct behaviour according to configuration ! (also if you see the language switcher URL you will see that default language does not have language code)


**Issue** #11143  ("Remove language code" for default language URL is OFF (lang-filter plugin)) 
- type or click the home page url without language code, you are redirected to home-page URL with language code of the last switched language

Please test without cache / page cache, or clean cache after every configuration change of language filter plugin

**Comment by ggppdk** [comment]
I have tested this item :white_check_mark: successfully on 963c58ccd4184232d3e88f0897666a60059341a3<br /><br />Works, see my previous comment, note i tested in browsers  (firefox, chrome, edge, IE)

NOTE: if you don't see the 301 redirect in network console when clicking on URL that is supposed to do 301 redirect, it is because some browser clear their console on new page load, and if localhost it may update too fast to see it<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by dgt41** [comment]
I have tested this item :white_check_mark: successfully on 963c58ccd4184232d3e88f0897666a60059341a3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by ggppdk** [comment]
at least 1 more tester for this ?

**Comment by Handy84** [comment]
I have tested this item :white_check_mark: successfully on https://github.com/joomla/joomla-cms/commit/963c58ccd4184232d3e88f0897666a60059341a3

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 963c58ccd4184232d3e88f0897666a60059341a3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by infograf768** [comment]
RTC and changed Title to add Regression. Thanks!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by infograf768** [comment]
@wilsonge 
Please get into 3.6.1

**Comment by infograf768** [comment]
Changing back to pending. I have issues here in Firefox when browser caching enabled (for Home page). Can't switch to default language (en-GB here).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by infograf768** [comment]
Folks, please test reverting
https://github.com/joomla/joomla-cms/pull/10344

It solved the issue here.

**Comment by ggppdk** [comment]
@infograf768 

Web-sites will not (at least they should not) change settings after initial setup for
- language filter plugin parameter "remove language code" 
- system cache plugin parameter "browser cache"

Otherwise the 301 , 304 etc instructions to the visitor browsers are no longer valid, and its not possible to force browser to clean cache for a specific URL that the browser does not resend to the server, but instead retrieves it from the cache

if you clear both Joomla cache and browser cache then it should work ?
- i tested and it works, maybe you can give more detailed instructions

at least for me it  worked (i cleared browser cache and Joomla cache after changing settings of the plugins)

**Comment by infograf768** [comment]
@ggppdk 
Yep, I know a bit how it should work. :)
I was successful when my Firefox had cache enabled when both:
Reverting #10344
Applying this patch

**Comment by ggppdk** [comment]
hhhmm i have it applied,

and i saw the effect you describe with SEF off, 
- but it worked after clean cache in chrome,

later today i will retest with firefox and other browsers

**Comment by infograf768** [comment]
I have tested this item :red_circle: unsuccessfully on 963c58ccd4184232d3e88f0897666a60059341a3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by chrisdavenport** [comment]
Why are we using a 301 here?  301 is a **permanent** redirect.  By telling caches not to cache a 301 you're effectively trying to convert it to a **temporary** redirect.  But there is already a 302 redirect for exactly that purpose. Telling caches not to cache a 302 makes perfect sense and you might have less trouble with caches as a result.

**Comment by ggppdk** [comment]
@chrisdavenport 

301, was decided to be re-applied, after some discussion for SEO reasons, in this PR
https://github.com/joomla/joomla-cms/pull/8894

I am not 100% sure it is worth it, 
but possibly 301 with zero expired time is the best choice ?

and then this PR is does the above, to fix problems related with 301 redirect
and it seems to almost succeed,

i will test the reported conflict of this fix with #10344

**Comment by chrisdavenport** [comment]
#8894 changed it from the default 303 to 301.  As @andrepereiradasilva pointed out, this changed the semantics to that of a permanent redirect, which you're now trying to override by adding cache directives instead of using the correct, temporary redirect in the first place.

303 will lead to duplicate content on search engines because, as the specification says, "the new URI in the Location header field is not considered equivalent to the effective request URI." (https://tools.ietf.org/html/rfc7231#section-6.4.4)

Adding the cache directives is good, but the status code should be changed too, in my opinion.

**Comment by 01Kuzma** [comment]
Which files are patched? If I change only the plugins/system/languagefilter/languagefilter.php Joomla doesn't load content at all.
When I could download the upgrade?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by andrepereiradasilva** [comment]
i agree with @chrisdavenport . i'm not enterily in favour of a "301 Permanent Redirect" in the language filter redirect as i explained in the original PR that made that change https://github.com/joomla/joomla-cms/pull/8894.

This PR only propose is to send HTTP headers for browsers to not cache language filter HTTP redirects (independently of the HTTP redirect type), doesn't changes the redirect HTTP status code type.

**Comment by Sandra97** [comment]
I have tested this item :red_circle: unsuccessfully on 963c58ccd4184232d3e88f0897666a60059341a3<br /><br />Able to reproduce the issue, I tested manually the patch, unsuccesfully. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@Sandra97 did you cleared the cookies and browser cache, or used a new private browser window for testing?

**Comment by Sandra97** [comment]
@andrepereiradasilva, I cleared the cookies and browser cache, tried with private browser and 2 different browsers but no change: 
joomla 3 6 - joomla 3 6 - ultime caratteristiche 2016-07-17 08-00-29
And all the menu items are now `/it`, some of the content displayed is in italian and some is english. If I revert the change, the display is ok, everything is nearly ok except of course that I can't display the default homepage in the right language

**Comment by andrepereiradasilva** [comment]
are you using todays latest staging?

remember this patch was merged yesterday https://github.com/joomla/joomla-cms/pull/10344 so when using the patchtester, since the patchtester replaces the entire file, appling this PR includes the changes from the merged PR in this file too.

So try to apply the changes manually please: https://github.com/joomla/joomla-cms/pull/11161/files

**Comment by Sandra97** [comment]
I did the test by replacing manually the current content of the file by yours (whole content)

**Comment by andrepereiradasilva** [comment]
> I did the test by replacing manually the current content of the file by yours (whole content)

You can't. that's what i'm saying, please add just the lines of code this PR adds.

**Comment by Sandra97** [comment]
Ok, I'll try like this.

**Comment by Sandra97** [comment]
I have tested this item :white_check_mark: successfully on 963c58ccd4184232d3e88f0897666a60059341a3<br /><br />Adding only the new lines to the file does the job. I'm now able to browse the default homepage.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by andrepereiradasilva** [comment]
thanks @Sandra97 

@infograf768 @ggppdk and other testers.

To test:
- clean cookies and browser cache before testing OR use new private window.
- and use latest staging from today + patch tester OR apply the patch manually by changing just the lines this patch changes (see https://github.com/joomla/joomla-cms/pull/11161/files).

**Comment by infograf768** [comment]
@andrepereiradasilva 
i tested on the latest staging with 10344. browser cache on and cleaned(which i don't do usually). no jomla cache. cookies cleared.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by andrepereiradasilva** [comment]
ok then try without 10344 applied, because this PR just tells the browser to not cache the redirect nothing more.

**Comment by infograf768** [comment]
As I said, to get it to work, I reverted 10344 and then applied your patch. And then it works here.

**Comment by andrepereiradasilva** [comment]
then change your test to success please

**Comment by infograf768** [comment]
I can't as #10344 is in the way for me here.

**Comment by mbabker** [comment]
The fact that 10344 needs to be reverted for this to work tells me that the plugin's constructor is doing more than setting application values, which is exactly what that pull request removed from the plugin to force the event dispatcher to correctly respect the configured plugin ordering values.  This also tells me that parts of the languagefilter plugin should NOT be the plugin but integrated directly into the application class because there is an extremely tight coupling to its constructor logic from within the application and this should NEVER be the case.

**Comment by infograf768** [comment]
@mbabker

I tested again after quitting Firefox and relaunching.
I don't have the issue anymore after this patch. Means I do not have anymore to revert 10344
I have the feeling my problem was due to sessions (not cookies).

**Comment by mbabker** [comment]
Well, that's a good sign.

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 963c58ccd4184232d3e88f0897666a60059341a3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by infograf768** [comment]
RTc again.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Comment by ggppdk** [comment]
@andrepereiradasilva 
@infograf768 

please read, or may i should open another issue

this patch seems good after cache clearing but there is another issue

e.g I have site with default language e.g. "French"

1- if current language is French and i visit    domain/ i get no redirection
2- if current language is German and i visit   domain/ i get a 301 redirection to domain/de/
3- if current language is English and i visit    domain/ i get a 301 redirection to domain/en/
4- if i visit    domain/fr/ i get a 301 redirection to domain/

1 and 4 are good
but 2 and 3 the redirection must not have code 301 ! , right ?

**Comment by ggppdk** [comment]
Different issue, i will open new issue

**Comment by andrepereiradasilva** [comment]
@ggppdk 

I think that's the correct behaviour if you have "Remove URL language code" set to Yes and your default language is French.

Try with this options.

image

**Comment by andrepereiradasilva** [comment]
deleted comment

**Comment by Bakual** [comment]
Imho, 301 is indeed the wrong redirect as the origin URL is still valid. It hasn't moved permanently. It's only temporary or "see other". So 302 or 303 are the appropriate redirects.
On a sidenote: google.com redirects using a 302 to the localised country domain (eg google.ch).

**Comment by nelgin** [comment]
I've been tearing my hair out trying to figure what I may have changed to cause the language to get stuck. I applied the patch, learned Joomla and browser cache and now it seems to be functioning as it should be.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11161">issues.joomla.org/joomla-cms/11161</a>.</sub>

**Pull request merged**